src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1057 @@
+            vaplic_emul_srccfg_access(acc);
+            break;
+        // case APLIC_MMSIADDRCFG_OFF:
+        //     vaplic_emul_mmsiaddrcfg_access(acc);
+        //     break;
+        // case APLIC_MMSIADDRCFGH_OFF:
+        //     vaplic_emul_mmsiaddrcfgh_access(acc);
+        //     break;
+        // case APLIC_SMSIADDRCFG_OFF:
+        //     vaplic_emul_smsiaddrcfg_access(acc);
+        //     break;
+        // case APLIC_SMSIADDRCFGH_OFF:
+        //     vaplic_emul_smsiaddrcfgh_access(acc);
+        //     break;
+        case APLIC_SETIP_OFF ... APLIC_SETIP_OFF+((APLIC_NUM_SETIx_REGS-1)*4):
+            vaplic_emul_setip_access(acc);
+            break;
+        case APLIC_SETIPNUM_OFF:
+            vaplic_emul_setipnum_access(acc);
+            break;
+        case APLIC_IN_CLRIP_OFF ... APLIC_IN_CLRIP_OFF+((APLIC_NUM_CLRIx_REGS-1)*4):
+            vaplic_emul_in_clrip_access(acc);
+            break;
+        case APLIC_CLRIPNUM_OFF:
+            vaplic_emul_clripnum_access(acc);
+            break;
+        case APLIC_SETIE_OFF ... APLIC_SETIE_OFF+((APLIC_NUM_SETIx_REGS-1)*4):
+            vaplic_emul_setie_access(acc);
+            break;
+        case APLIC_SETIENUM_OFF:
+            vaplic_emul_setienum_access(acc);
+            break;
+        case APLIC_CLRIE_OFF ... APLIC_CLRIE_OFF+((APLIC_NUM_CLRIx_REGS-1)*4):
+            vaplic_emul_clrie_access(acc);
+            break;
+        case APLIC_CLRIENUM_OFF:
+            vaplic_emul_clrienum_access(acc);
+            break;
+        // case APLIC_SETIPNUM_LE_OFF:
+        //     vaplic_emul_setipnum_le_access(acc);
+        //     break;
+        // case APLIC_SETIPNUM_BE_OFF:
+        //     vaplic_emul_setipnum_be_access(acc);
+        //     break;
+        // case APLIC_GENMSI_OFF:
+        //     vaplic_emul_genmsi_access(acc);
+        //     break;
+        case APLIC_TARGET_OFF ...APLIC_TARGET_OFF+((APLIC_MAX_INTERRUPTS-2)*4):
+            vaplic_emul_target_access(acc);
+            break;
+        default:
+            if(!acc->write) {
+                vcpu_writereg(cpu()->vcpu, acc->reg, 0);
+            }
+            break;
+    }
+
+    return true;
+}
+
+/**
+ * @brief Function to handle writes (or reads) to (from) IDC structure.
+ * 
+ * @param acc emulated access
+ * @return true  if conclude without errors.
+ * @return false if the access is not aligned.
+ */
+static bool vaplic_idc_emul_handler(struct emul_access *acc)
+{
+    // only allow aligned word accesses
+    if (acc->width > 4 || acc->addr & 0x3) return false;
+
+    int idc_id = ((acc->addr - APLIC_IDC_OFF) >> 5) & 0x3ff;
+    if(!(idc_id < cpu()->vcpu->vm->arch.virqc.idc_num)){
+        if(!acc->write) {
+            vcpu_writereg(cpu()->vcpu, acc->reg, 0);
+        }
+        return true;
+    }
+    uint32_t addr = acc->addr - platform.arch.plic_base + APLIC_IDC_OFF;
+    addr = addr - (sizeof(irqc_hart[0]) * idc_id);
+    switch (addr & 0x1F) {
+        case APLIC_IDC_IDELIVERY_OFF:
```

Comment:
Similar to the comment above, this code looks like this:

```
case offsetof(struct aplic_hart_hw, idelivery):
            vaplic_emul_idelivery_access(acc, idc_id);
            break;
```
This will allow us to remove the `#defines` macro OFF from aplic.h